web/docs/guides/middleware-customization.md

```diff
@@ -0,0 +1,134 @@
+- [express.urlencoded](https://expressjs.com/en/api.html#express.urlencoded) / [body-parser](https://expressjs.com/en/resources/middleware/body-parser.html#bodyparserurlencodedoptions): Returns middleware that only parses urlencoded bodies and only looks at requests where the `Content-Type` header matches the type option.
+- [cookieParser](https://github.com/expressjs/cookie-parser#readme): Parse Cookie header and populate `req.cookies` with an object keyed by the cookie names.
+
+## Customization
+
+You have three places where you can customize middleware:
+1. global: here, any changes will apply by default *to all operations (`query` and `action`) and `api`.* This is helpful if you wanted to add support for multiple domains to CORS, for example. ⚠️ Please treat modifications to global middleware with extreme care!
+2. per-api: you can override middleware for a specific api route (exe: `POST /webhook/callback`). This is helpful if you want to disable JSON parsing, for example.
+3. per-path: this is helpful if you need to customize middleware for all methods for a given path. This is helpful for things like "complex CORS requests," which may need to apply to both `OPTIONS` and `GET`, or to apply some middleware to a _set of `api` routes_.
+
+### Types
+
+Below are the relevant TS types and the actual definitions of default middleware.
+
+```ts
+export type MiddlewareConfig = Map<string, express.RequestHandler>
+
+export type MiddlewareConfigFn = (middleware: MiddlewareConfig) => MiddlewareConfig
+
+const defaultGlobalMiddleware: MiddlewareConfig = new Map([
+  ['helmet', helmet()],
+  ['cors', cors({ origin: config.allowedCORSOrigins })],
+  ['logger', logger('dev')],
+  ['express.json', express.json()],
+  ['express.urlencoded', express.urlencoded({ extended: false })],
+  ['cookieParser', cookieParser()]
+])
+```
+
+Here is the [definition of config](https://github.com/wasp-lang/wasp/blob/main/waspc/data/Generator/templates/server/src/config.js).
+
+## 1. Customize global middleware
+
+If you would like to modify the middleware for _all_ operations and APIs, you can do something like:
+
+```c title=todoApp.wasp
+app todoApp {
+  // ...
+
+  server: {
+    setupFn: import setup from "@server/serverSetup.js",
+    middlewareConfigFn: import { serverMiddlewareFn } from "@server/serverSetup.js"
+  },
+}
+```
+
+```ts title=src/server/serverSetup.js
+import cors from 'cors'
+import { MiddlewareConfigFn } from '@wasp/middleware'
+import config from '@wasp/config.js'
+
+export const serverMiddlewareFn: MiddlewareConfigFn = (middleware) => {
+  // Example of adding an extra domains to CORS.
+  middleware.set('cors', cors({ origin: [config.frontendUrl, 'https://example1.com', 'https://example2.com'] }))
```

Comment:
Aha, making config part of API here. Hmmmm. This asks a good question, what is part of API? We don't really have a way to track it, we kind of know it, but we don't have it "secured" in a code -> I could fiddle with `config.js` and forget we promised users it will be there!
In general yeah I think we can expose it, it is a way to get to env vars. Although they don't know much about config, we would have to tell them what is in it then, how would they know it has `frontendUrl`?. Uff, ok this gets complex now hm. Maybe we can leave it a bit greyish for now? Or do you have a better idea how to go about it?